docs/v2_design.md

```diff
@@ -2,188 +2,600 @@
+
+Such design unifies two ways to construct a quantity:
+- through initialization of its quantity type
+
+  ```cpp
+  quantity<isq::speed[km / h]> speed{60.};
+  ```
+
+- via a multiplication of a number and a quantity reference
+
+  ```cpp
+  auto speed = 60. * isq::speed[km / h];
+  ```
+
+### Named quantity and named unit
+
+```cpp
+using namespace mp_units;
+auto p = quantity<isq::power[si::watt], int>{5};
+```
+
+or
+
+```cpp
+using namespace mp_units;
+auto p = 5 * isq::power[si::watt];
+```
+
+or
+
+```cpp
+using namespace mp_units;
+using namespace mp_units::si::unit_symbols;
+auto p = 5 * isq::power[W];
+```
+
+```cpp
+static_assert(std::is_same_v<decltype(p), quantity<reference<isq::power, si::watt>{}, int>>);
+static_assert(std::is_same_v<decltype(p.dimension), derived_dimension<power<isq::dim_length, 2>, isq::dim_mass, per<power<isq::dim_time, 3>>>>);
+static_assert(p.quantity_spec != isq::force * isq::speed);
+static_assert(interconvertible(p.quantity_spec, isq::force * isq::speed));
+static_assert(p.dimension == isq::force.dimension * isq::speed.dimension);
+```
 
-### Quantity Comparison
+### Named quantity and derived (unnamed) unit
+
+```cpp
+using namespace mp_units;
+using namespace mp_units::si::unit_symbols;
+auto s = quantity<isq::speed[m / s], int>{5};
+```
+
+or
+
+```cpp
+using namespace mp_units;
+auto s = 5 * isq::speed[si::metre / si::second];
+```
+
+or
+
+```cpp
+using namespace mp_units;
+using namespace mp_units::si::unit_symbols;
+auto s = 5 * isq::speed[m / s];
+```
+
+```cpp
+static_assert(std::is_same_v<decltype(s),
+              quantity<reference<isq::speed, derived_unit<si::metre, per<si::second>>{}>{}, int>>);
+```
+
+
+## Concepts
+
+### `quantity_of<T>`
+
+Checks whether the quantity is exactly of a specific property where property can be:
+- `quantity_spec`
+- dimension
+- `reference`
+
+```cpp
+static_assert(quantity_of<decltype(10 * isq::speed[m/s] / (2 * isq::length[m]) * (4 * isq::time[s])), dimension_one>);
+static_assert(!quantity_of<decltype(10 * isq::length[m]), isq::height>);   // kind of
+static_assert(!quantity_of<decltype(10 * isq::width[m]), isq::height>);    // different kinds
+static_assert(quantity_of<decltype(10 * isq::speed[m/s]), isq::speed>);
+static_assert(!quantity_of<decltype(20 * isq::length[m] / (2 * isq::time[s])), isq::speed>);  // derived unnamed quantity
+static_assert(quantity_of<decltype(10 * isq::speed[m/s]), isq::speed[m /s]>);
+static_assert(!quantity_of<decltype(10 * isq::speed[m/s]), isq::speed[km / h]>);  // different unit
+```
+
+
+### `weak_quantity_of<T>`
+
+Checks whether the quantity is of an interconvertible `quantity_spec` or of the same dimension
+and unit (in case of `reference`). It will fail for non-equivalent but inconvertible unit.
+
+```cpp
+static_assert(weak_quantity_of<decltype(10 * isq::speed[m/s] / (2 * isq::length[m]) * (4 * isq::time[s])), dimension_one>);
+static_assert(weak_quantity_of<decltype(10 * isq::length[m]), isq::height>);   // kind of
```

Comment:
"Kind of" doesn't match the order of the arguments. Length is not a kind of height, but the other way around.

Reply:
I mean, if you consider that this operation stands for "kind of", as `{ length } -> kind_of<height>` in a concept, that's wrong.

Reply:
Actually, it works in both cases here. As long as both types are on the same branch of the tree of kinds, they will make this concept happy. Maybe a comment should be replaced with better text. I just couldn't find a better hint.

Reply:
Maybe "convertible kinds", if you allow converting height to length.